gobblin-yarn/src/main/java/gobblin/yarn/GobblinWorkUnitRunner.java

```diff
@@ -0,0 +1,413 @@
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.yarn.api.ApplicationConstants;
+import org.apache.hadoop.yarn.api.records.ContainerId;
+import org.apache.hadoop.yarn.util.ConverterUtils;
+import org.apache.helix.HelixDataAccessor;
+import org.apache.helix.HelixManager;
+import org.apache.helix.HelixManagerFactory;
+import org.apache.helix.HelixProperty;
+import org.apache.helix.InstanceType;
+import org.apache.helix.NotificationContext;
+import org.apache.helix.api.id.StateModelDefId;
+import org.apache.helix.messaging.handling.HelixTaskResult;
+import org.apache.helix.messaging.handling.MessageHandler;
+import org.apache.helix.messaging.handling.MessageHandlerFactory;
+import org.apache.helix.model.Message;
+import org.apache.helix.task.TaskFactory;
+import org.apache.helix.task.TaskStateModelFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.JmxReporter;
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.MetricSet;
+import com.codahale.metrics.jvm.FileDescriptorRatioGauge;
+import com.codahale.metrics.jvm.GarbageCollectorMetricSet;
+import com.codahale.metrics.jvm.MemoryUsageGaugeSet;
+import com.codahale.metrics.jvm.ThreadStatesGaugeSet;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.eventbus.EventBus;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.Service;
+import com.google.common.util.concurrent.ServiceManager;
+
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.runtime.TaskExecutor;
+import gobblin.runtime.TaskStateTracker;
+import gobblin.yarn.event.DelegationTokenUpdatedEvent;
+
+
+/**
+ * The main class running in the containers managing services for running Gobblin
+ * {@link gobblin.source.workunit.WorkUnit}s.
+ *
+ * <p>
+ *   This class serves as a Helix participant and it uses a {@link HelixManager} to work with Helix.
+ *   This class also uses the Helix task execution framework and {@link GobblinHelixTaskFactory} class
+ *   for creating {@link GobblinHelixTask}s that Helix manages to run Gobblin data ingestion tasks.
+ * </p>
+ *
+ * @author ynli
+ */
+public class GobblinWorkUnitRunner {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(GobblinWorkUnitRunner.class);
+
+  static final String GOBBLIN_TASK_FACTORY_NAME = "GobblinTaskFactory";
+
+  private final ContainerId containerId;
+
+  private final HelixManager helixManager;
+
+  private final ServiceManager serviceManager;
+
+  private final EventBus eventBus;
+
+  private final TaskStateModelFactory taskStateModelFactory;
+
+  private final MetricRegistry metricRegistry = new MetricRegistry();
```

Comment:
If possible, I think it may be better to just use the Gobblin Metrics API:

* I think it makes the code more consistent if we stick to just using Gobblin Metrics - only the gobblin-metrics submodule should be aware that we are using the codahale library - which also makes migration to a different metrics backend easier.
* Gobblin Metrics lets us take advantage of the child-parent aggregation
* Gobblin Metrics allows us to report metrics to Kafka also
* Gobblin Metrics has EventReporters that may be useful throughout the code